### PR TITLE
add test framework for doing full request webdav tests

### PR DESCRIPTION
**Pull request by icewind1991**
Makes it easy to write test that emulate a full webdav request, including all the sabre logic.

The tests are initially meant for #17811 but I figured it would be better to review this seperately.

cc @DeepDiver1975 @PVince81

### CODE REVIEW

File: tests/lib/connector/sabre/requesttest/requesttest.php [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+	}
+
+	/**
+	 * @param View $view
+	 * @param string $user
+	 * @param string $password
+	 * @param ExceptionPlugin $exceptionPlugin
+	 * @return Server
+	 */
+	protected function getSabreServer(View $view, $user, $password, ExceptionPlugin $exceptionPlugin) {
```
Comment (PVince81) [inline tests/lib/connector/sabre/requesttest/requesttest.php]: Please move this into a real class instead of duplicating the init code.
Then use that method also from remote.php.

**Comment by PVince81** [comment]
Otherwise this is a nice addition. :smile:

File: tests/lib/connector/sabre/requesttest/requesttest.php (hunk 1 above)
Comment (icewind1991) [inline tests/lib/connector/sabre/requesttest/requesttest.php]: Done

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by icewind1991** [comment]
jenkins passed on ci-pr

**Comment by icewind1991** [comment]
@DeepDiver1975 @PVince81 @MorrisJobke can this be merged?

File: apps/files/appinfo/remote.php [hunk 2]
```diff
@@ -33,51 +33,23 @@
 // Turn off output buffering to prevent memory problems
 \OC_Util::obEnd();
 
+$serverFactory = new \OC\Connector\Sabre\ServerFactory(
```
Comment (PVince81) [inline apps/files/appinfo/remote.php]: Awwwyeah

File: apps/files_sharing/publicwebdav.php [hunk 3]
```diff
@@ -33,24 +33,18 @@
 // Backends
 $authBackend = new OCA\Files_Sharing\Connector\PublicAuth(\OC::$server->getConfig());
 
-// Fire up server
-$objectTree = new \OC\Connector\Sabre\ObjectTree();
-$server = new \OC\Connector\Sabre\Server($objectTree);
-// Set URL explicitly due to reverse-proxy situations
-$server->httpRequest->setUrl(\OC::$server->getRequest()->getRequestUri());
-$server->setBaseUri($baseuri);
+$serverFactory = new \OC\Connector\Sabre\ServerFactory(
```
Comment (PVince81) [inline apps/files_sharing/publicwebdav.php]: Did you exclude the tags plugin ? Public webdav doesn't have exactly the same plugins as in regular webdav. Please double check.

Comment (PVince81) [inline apps/files_sharing/publicwebdav.php]: Best is to register "additional non-base plugins" in the callback.

Comment (DeepDiver1975) [inline apps/files_sharing/publicwebdav.php]: may I suggest to move the code from remote.php to its own class - just like here:
https://github.com/DeepDiver1975/dav/blob/master/appinfo/remote.php

Comment (DeepDiver1975) [inline apps/files_sharing/publicwebdav.php]: this class can be easily reused in other situations ...

Comment (icewind1991) [inline apps/files_sharing/publicwebdav.php]: > Did you exclude the tags plugin ? Public webdav doesn't have exactly the same plugins as in regular webdav. Please double check.

The tags plugin is only registered when a user is loggedin.

As for any further refactoring, I would prefer to do those in a separate PR

File: lib/private/connector/sabre/serverfactory.php [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+			// Create ownCloud Dir
+			$rootDir = new \OC\Connector\Sabre\Directory($view, $rootInfo);
+			$objectTree->init($rootDir, $view, $this->mountManager);
+
+			$server->addPlugin(new \OC\Connector\Sabre\TagsPlugin($objectTree, $this->tagManager));
+			$server->addPlugin(new \OC\Connector\Sabre\QuotaPlugin($view));
+
+			// custom properties plugin must be the last one
+			if($this->userSession->isLoggedIn()) {
```
Comment (PVince81) [inline lib/private/connector/sabre/serverfactory.php]: From what I see here only CustomPropertiesBackend is loaded when the user is logged in, so you'd need to move TagsPlugin into that block too

Comment (icewind1991) [inline lib/private/connector/sabre/serverfactory.php]: done

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

File: lib/private/connector/sabre/serverfactory.php [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+		// Fire up server
+		$objectTree = new \OC\Connector\Sabre\ObjectTree();
+		$server = new \OC\Connector\Sabre\Server($objectTree);
+		// Set URL explicitly due to reverse-proxy situations
+		$server->httpRequest->setUrl($requestUri);
+		$server->setBaseUri($baseUri);
+
+		// Load plugins
+		$defaults = new \OC_Defaults();
+		$server->addPlugin(new \OC\Connector\Sabre\BlockLegacyClientPlugin($this->config));
```
Comment (PVince81) [inline lib/private/connector/sabre/serverfactory.php]: When comparing the lists, this one also isn't present in publicwebdav. Let's hope this won't cause side effects.

**Comment by PVince81** [comment]
- [x] TEST: simple sync
- [ ] TEST: smashbox basic sync
- [ ] TEST: federation folder mount
- [ ] TEST: federation file mount
- [ ] TEST: folder public link
- [ ] TEST: file public link

**Comment by PVince81** [comment]
- [ ] BUG: federation file link appears as folder

@icewind1991 please have a look

**Comment by icewind1991** [comment]
> BUG: federation file link appears as folder

Fixes

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by PVince81** [comment]
:+1: works

**Comment by DeepDiver1975** [comment]
:+1:
